Improve MOTD documentation (#2534)
canonical/ubuntu-pro-client: docs/explanations/motd_messages.md

When the Ubuntu Pro Client (`pro`) is installed on the system, it delivers
custom messages on ["Message of the Day" (MOTD)](https://wiki.debian.org/motd).
Those messages are generated directly by three different sources.

> 

Reviewer: Once we're done tweaking the doc, could we add a bullet list here with the three sources? (I know the doc isn't long, but it's a helpful thing to have)

Author: ack and done

> ## MOTD notifying about available updates

Reviewer: ```suggestion
## MOTD messages about available updates
```

Reviewer: "MOTD message" or "MOTD messages" is more likely to be a search term, I think

The [update-notifier](https://wiki.ubuntu.com/UpdateNotifier) delivers a script
via the `update-notifier-common` package called
`/usr/lib/update-notifier/apt_check.py.
With regards to Ubuntu Pro, this script is responsible for:
  

> * Informing the user about the status of one of the ESM services; `esm-apps` if

Reviewer: Can it be both?

  the machine is an LTS series, or `esm-infra` if the series is in ESM mode.
* Showing the number of `esm-infra` or `esm-apps` packages that can be upgraded
  on the machine.

> `update-notifier` always was responsible to add information about potential

Reviewer: ```suggestion
`update-notifier` has always been used to add information about potential
```

Author: has always added

> updates to MOTD to raise user awareness, with the existence of pro they are

Reviewer: ```suggestion
updates to MOTD to raise user awareness. With the advent of Ubuntu Pro they are
```

just more differentiated.

Note that if you run `apt_check.py` directly it might give you rather

> unreadable output as it is meant for program use, you can add `--human-readable`

Reviewer: ```suggestion
unreadable output as it is meant for program use. You can add `--human-readable`
```

> to get the same info you see in MOTD.

Reviewer: Implied that the information you see in MOTD is different when using the `--human-readable` flag

### Machine is unattached

> On a machine that runs a Ubuntu release for which the `esm-apps` service

Reviewer: ```suggestion
On a machine that runs an Ubuntu release for which the `esm-apps` service
```

> is available, but not yet attached to Ubuntu Pro in any way there will

Reviewer: ```suggestion
is available, but not yet attached to an Ubuntu Pro subscription, there will
```

> only be a hint that there could be more available through ESM.

Reviewer: ```suggestion
be a message notifying the user that there may be more security updates available through ESM.
```

Reviewer: Although "hint" isn't wrong, it may have some connotations that we don't intend to imply. I think message would generally be clearer and have less room for interpretation.

```
Expanded Security Maintenance for Applications is not enabled.

0 updates can be applied immediately.

Enable ESM Apps to receive additional future security updates.
See https://ubuntu.com/esm or run: sudo pro status
```

### Machine is fully attached

> Then as the opposite example, here the output users will see in MOTD
> when run on a LTS machine with `esm-infra` and `esm-apps` services enabled:

Reviewer: ```suggestion
In the opposite situation, if an LTS machine has the `esm-infra` and `esm-apps` services enabled then users will see the following output in MOTD:
```

When the Ubuntu Pro Client (`pro`) is installed on the system, it delivers
custom messages on ["Message of the Day" (MOTD)](https://wiki.debian.org/motd).
Those messages are generated directly by three different sources.

* MOTD about available updates
* MOTD about important subscription conditions
* MOTD about ESM being available

## MOTD about available updates

The [update-notifier](https://wiki.ubuntu.com/UpdateNotifier) delivers a script
via the `update-notifier-common` package called
`/usr/lib/update-notifier/apt_check.py.
With regards to Ubuntu Pro, this script is responsible for:

* Informing the user about the status of one of the ESM services; `esm-apps` if
  the machine is an LTS series, or `esm-infra` if the series is in ESM mode.
* Showing the number of `esm-infra` or `esm-apps` packages that can be upgraded
  on the machine.

> `update-notifier` always was responsible to add information about potential

Reviewer: For some reason it left this line in here, so it's rendering both this and the following one (the one we edited it to)

To see these additional updates run: apt list --upgradable
```

> ### Machine is fully attached, but already in ESM

Reviewer: I got a bit confused by the wording in this section. I wasn't sure why, if the machine is attached and in ESM, why only ESM infra being advertised. I asked about this earlier and it was confirmed that whichever services are enabled get shown but I'm not sure that's clear here. We probably want to revise this wording to make sure it reflects what's currently happening after all the changes to terminology.

(I should clarify, I attached the comment to the sub-heading and not the paragraph below it by mistake, sorry!)

However, if we were running this on a Ubuntu release that has already
entered ["ESM"](https://ubuntu.com/security/esm), we would instead see

> `esm-infra` being advertised:

Author: ack

To see these additional updates run: apt list --upgradable
```

### Machine is fully attached, on an older release

Above you have seen examples of recent (as in "still in their first 5

> years of support") ubuntu releases, where the hint is about ESM Apps

Reviewer: ```suggestion
years of support") Ubuntu releases, where the hint is about ESM Apps
```

Author: done

> extending the coverage to the universe repositories.

Reviewer: ```suggestion
extending the coverage to the Universe repositories.
```

Author: done

> However, if running on an Ubuntu release that has is already past the initial

Reviewer: This whole section is much clearer now, thank you!

information for `esm-infra` if the service was disabled and the series running
on the machine is in ESM state.

## MOTD about important subscription conditions

One of the [timer jobs](https://canonical-ubuntu-pro-client.readthedocs-hosted.com/en/latest/explanations/what_are_the_timer_jobs.html)
Ubuntu Pro uses can insert additional messages into MOTD.

> These messages will be always delivered close to the content created by

Reviewer: I'm not sure what "close to" means in this context - is there a more specific word that we can use instead?

Author: yep, "... delivered next to the content created by `update-notifier`."
Done

Your grace period will expire in 9 days.
```

## MOTD for the ESM announcement

> There was also a general announcement info that further helped to raise
> awareness of Pro being available and for free. This was only shown on systems
> covered by `esm-apps`. Furthermore it was meant to be only shown for a limited
> time and therefore removed in 27.14.

Reviewer: ```suggestion
When Ubuntu Pro became generally available, a temporary announcement was made
through MOTD. This was intended to raise awareness of Pro now being available
and free for personal use, and was shown on systems covered by `esm-apps`. 
```

Author: systems that could be covered by

Author: done

It looked like:

```
 * Introducing Expanded Security Maintenance for Applications.
   Receive updates to over 25,000 software packages with your
   Ubuntu Pro subscription. Free for personal use.

     https://ubuntu.com/pro
```

> 

Reviewer: Since this message was intended as a limited-time announcement to coincide
with the release of Ubuntu Pro into general availability, it was removed in
27.14.

Author: done

> ## How are these messages inserted into MOTD and how could I disable them?

Reviewer: ```suggestion
## How are these messages inserted into MOTD and how can I disable them?
```

Reviewer: "can" is more likely to be searched for

> Just like there are different aspects to the messages outlined above they
> come from different sources into the MOTD that one sees at login time.

Reviewer: ```suggestion
Just as there are different purposes to the messages outlined above,
there are different sources producing the MOTD that one sees at login.
```

Author: the MOTD -> "these MOTD elements"

Those messages are considered important to ensure user awareness about
the free additional security coverage provided by Ubuntu Pro and about
yet unapplied potential updates in general. Therefore it is not recommended

> to disable them, but if one wants to do that anyway they can be disabled
> selectively via the config files that add them as outlined below.

Reviewer: ```suggestion
to disable them, but you can selectively disable them via the config
files that add them, as outlined below.
```

> Removing those is considered a conffile change to customize a program

Reviewer: I'm not sure what "those" are being referred to here - please clarify

and they will stay removed even on future upgrades or re-installations of the

> related packages. If you realize that you actually need them back you need
> to reinstall the related packages and tell apt/dpkg to offer you to restore
> those files via:
> `sudo apt install --reinstall -o Dpkg::Options::="--force-confask" ubuntu-advantage-tools update-notifier-common`

Reviewer: ```suggestion
related packages. If you later realise that you want them back, you will need
to reinstall the related packages and tell `apt`/`dpkg` to offer you to restore
those files using the following command:

```
sudo apt install --reinstall -o Dpkg::Options::="--force-confask" ubuntu-advantage-tools update-notifier-common```
```

Reviewer: Well...that didn't work as successfully as I'd hoped :P

Reviewer: Not sure why it wouldn't let me add a code block there, but for the sake of tidiness I like to have code snippets on a separate line (especially if they're long snippets)

> ## Source: MOTD notifying about available updates

Reviewer: If we change the header at the top, this should also be changed to match

1. `update-notifier-common` has a hook `/etc/apt/apt.conf.d/99update-notifier` that runs after `apt update`.
2. That hook will update the information in `/var/lib/update-notifier/updates-available` matching the new package information that was just fetched by using `/usr/lib/update-notifier/apt-check --human-readable`.
3. At MOTD generation time, the script located at `/etc/update-motd.d/90-updates-available` checks if `/var/lib/update-notifier/updates-available` exists and if it does, inserts the message into the full MOTD.

If you want to disable any message of update-notifier (not just related to Ubuntu Pro and ESM) about potentially available updates remove `/etc/update-motd.d/90-updates-available`.

## Source: MOTD for important contract conditions

> 1. The contract status is checked periodically in the background when the machine is attached to an Ubuntu Pro contract.

Author: => subscription
